File: ios/Expenzez/WidgetBridge.m
//
//  WidgetBridge.m
//  Expenzez
//
//  Objective-C glue that exposes the Swift WidgetBridge module to the React
//  Native bridge.
//
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(WidgetBridge, NSObject)

RCT_EXTERN_METHOD(setSnapshot:(NSString *)json
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(reloadAll:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/Expenzez/WidgetBridge.swift
//
//  WidgetBridge.swift
//  Expenzez
//
//  Native module that lets the React Native app push the widget snapshot into
//  the shared App Group store and ask WidgetKit to re-render the home-screen
//  widgets. Paired with WidgetBridge.m (RCT_EXTERN_MODULE) and consumed from JS
//  via NativeModules.WidgetBridge (see services/widget/widgetBridge.ts).
//
import Foundation
import WidgetKit

@objc(WidgetBridge)
class WidgetBridge: NSObject {

  // Must match the App Group on both the app and the widget extension, and the
  // key the widget reads in ExpenzezWidgets/WidgetSnapshot.swift.
  private let appGroup = "group.com.biszaal.expenzez"
  private let snapshotKey = "widget_snapshot"

  /// Persist the snapshot JSON into the App Group's UserDefaults.
  @objc(setSnapshot:resolver:rejecter:)
  func setSnapshot(_ json: String,
                   resolver resolve: RCTPromiseResolveBlock,
                   rejecter reject: RCTPromiseRejectBlock) {
    guard let defaults = UserDefaults(suiteName: appGroup) else {
      reject("no_app_group", "App Group \(appGroup) is unavailable", nil)
      return
    }
    defaults.set(json, forKey: snapshotKey)
    resolve(nil)
  }

  /// Ask WidgetKit to reload every widget timeline (no-op before iOS 14).
  @objc(reloadAll:rejecter:)
  func reloadAll(_ resolve: RCTPromiseResolveBlock,
                 rejecter reject: RCTPromiseRejectBlock) {
    if #available(iOS 14.0, *) {
      WidgetCenter.shared.reloadAllTimelines()
    }
    resolve(nil)
  }

  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
